### ISSUE TITLE
[feature request] Add multi-index, dim-as-sequence support to all tensor reduction operations

### ISSUE THREAD

**Issue by elistevens** [pytorch/pytorch#12316]
## 🚀 Feature

The `dim` parameter for `torch.sum` and e.g. `torch.max` do not have the same set of supported input types.  From https://pytorch.org/docs/master/torch.html#reduction-ops :

> torch.max(input, dim, keepdim=False, out=None) -> (Tensor, LongTensor)
> Parameters: dim (int) – the dimension to reduce

vs.

> torch.sum(input, dim, keepdim=False, dtype=None) -> Tensor
> Parameters: dim (int or tuple of python:ints) – the dimension or dimensions to reduce

This means that reducing a 5D tensor looks like one of the following:

    t.sum([2,3,4])
    t.view(t.size(0), t.size(1), -1).max(dim=2)[0]

And it would be nice to have the same style available for all reduction operations.

## Motivation

Having to spell similar operations in very different ways results in less readable code. It's harder to verify visually that the two operations do the equivalent thing.

## Pitch

I would like the multi-index `dim` support for `torch.sum` added to `torch.min`, `torch.max`, etc.

## Alternatives

The status quo works for my use case, but it's unclear if that remains true for users that need the second item of the `torch.max` return tuple.

**Comment by zou3519** [comment]
This should be coming soon to tensor iterators with @colesbury 's future changes.

**Comment by fmassa** [comment]
While with TensorIterator we can easily have the max / min implemented, getting the argmax/argmin is another story

**Comment by ngimel** [comment]
Closing, `torch.sum`, `torch.amin`, `torch.amax` support multi-dimensional reduction

**Comment by carlosgmartin** [comment]
@ngimel What about `softmax` and `log_softmax`, as @vadimkantorov asked?
